plugins/consensual_youtube.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/python -B
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#
+# consensual_youtube.py -- Pelican plugin that supports postponing loading
+# youtube resources until the user explicitly opts into playing the video
+#
+from os import path
+
+from urllib import request
+
+from pelican import contents, signals
+
+from bs4 import BeautifulSoup
+
+content_path = ''
+output_path = ''
+
+def pelican_init(pelicanobj):
+    global output_path
+
+    content_path = pelicanobj.settings['PATH']
+    output_path = pelicanobj.settings['OUTPUT_PATH']
+
+def generate_youtube(content):
+    global output_path
+
+    if isinstance(content, contents.Static):
+        return
+    soup = BeautifulSoup(content._content, 'html.parser')
+    tags = soup.find_all('youtube')
+    if tags:
+        for tag in tags:
+            tag.name = 'div';
+            yt_id = tag['youtube_id']
+            assert yt_id 
+
+            preview = f'/img/yt_preview_{yt_id}.jpg'
+            if not path.isfile(content_path + preview):
+                request.urlretrieve(f'https://img.youtube.com/vi/{yt_id}/0.jpg', output_path + preview)
```

Comment:
This is intentional: if the file is present in the 'content' input, pelican will copy it to the output. If not, the plugin downloads it to the output.

This way we don't have to make assumptions about ordering: it will work regardless of whether this plugin runs first or the pelican code that copies input resources to the output runs first.

You could argue for downloading it to the 'content' input directory. Depending on the ordering that might not work the first time, but it would encourage people to check in the image so it doesn't need to be re-downloaded on subsequent generations. I don't have a strong preference, happy to download it to `content_path + preview` if you think that is clearly better.